src/main/scala/com/fulcrumgenomics/bam/AnnotateMolecularBarcodes.scala

```diff
@@ -0,0 +1,148 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2016 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.bam
+
+import com.fulcrumgenomics.cmdline.{ClpGroups, FgBioTool}
+import com.fulcrumgenomics.util.ReadSegment.{MolecularBarcode, ReadSegment, Template}
+import com.fulcrumgenomics.util.ReadStructure
+import dagr.commons.CommonsDef.PathToBam
+import dagr.commons.io.Io
+import dagr.commons.util.LazyLogging
+import dagr.sopt._
+import dagr.sopt.cmdline.ValidationException
+import htsjdk.samtools._
+import htsjdk.samtools.util._
+
+import scala.collection.JavaConversions._
+
+@clp(description =
+  """
+    |Annotates reads with their molecular barcodes.  Only template bases will be retained as specified by the read
+    |structure.  If only one molecular barcode tag is given, then the molecular barcodes will be concatenated and stored
+    |in that tag.  Otherwise the number of molecular barcodes should match the number of tags.  Mapping information will
+    |be unchanged, as such, this tool should not be used on reads that have been mapped since it would lead to an BAM
+    |with inconsistent records.
+  """,
+  group = ClpGroups.SamOrBam)
+class AnnotateMolecularBarcodes
+( @arg(flag = "i",   doc = "Input BAM file.")                                     val input: PathToBam,
+  @arg(flag = "o",   doc = "Output BAM file.")                                    val output: PathToBam,
+  @arg(flag = "rs1", doc = "The read structure for read 1.")                       val readStructure1: String,
+  @arg(flag = "rs2", doc = "The read structure for read 2.")                       val readStructure2: String,
+  @arg(flag = "b1",  doc = "SAM tags in which to store the molecular barcodes (one-per segment) for read.")
```

Comment:
I personally think it would be nicer to supply just a single list of tags, with the understanding that UMIs are extracted from R1 first then R2 and matched to tags in that order.  Then you could also just call this `tags`.